File: algorithms/1_easy/C/03258_count_substrings_that_satisfy_k_constraint_i/sliding_window.cpp
class Solution {
public:
    int countKConstraintSubstrings(string s, int k) {
        const size_t len {s.size()};
        size_t left {0};
        size_t count {0};
        int zeros {0};
        int ones {0};

        for (size_t right {0}; right < len; ++right) {
            if (s[right] == '0') {
                zeros += 1;
            } else {
                ones += 1;
            }

            while (ones > k && zeros > k) {
                if (s[left] == '0') {
                    zeros -= 1;
                } else {
                    ones -= 1;
                }

                left += 1;
            }

            count += right - left + 1;
        }

        return static_cast<int>(count);
    }
};

File: algorithms/1_easy/C/03258_count_substrings_that_satisfy_k_constraint_i/sliding_window.py
class Solution:
    def countKConstraintSubstrings(self, s: str, k: int) -> int:
        n = len(s)
        left = 0
        count = 0
        zeros = 0
        ones = 0

        for right in range(n):
            if s[right] == '0':
                zeros += 1
            else:
                ones += 1

            while ones > k and zeros > k:
                if s[left] == '0':
                    zeros -= 1
                else:
                    ones -= 1
                left += 1

            count += right - left + 1

        return count
